### Title: 
# Exploratory Data Analysis

### Description:
We are going to perform an EDA of our dataset.

### Authors:
#### Hugo Cesar Octavio del Sueldo
#### Jose Lopez Galdon

### Date:
23/10/2020

### Version:
1.0

***

### Libraries

We are going to start this EDA loading the respective libraries

In [1]:
    # Numpy & Pandas to work with the DF
import numpy as np
import pandas as pd

    # Seaborn / matplotlib for graphs
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline

    # Visualize DF
from IPython.display import display, HTML

***

## Load data

In [2]:
    # To automate the work as much as possible, we will parameterize the codes, so in this case, we will create an objetct with
    # the path root
name = 'data_train_merge'

data = pd.read_csv(f'../data/02_intermediate/{name}.csv',  # Path root: here we include an f-string with the variable name to parametrize
                   low_memory = False)                     # To avoid warnings we use set low_memory = False

***

## Variable predict

Taking into account the goals of the practice, we only want to know those persons that `Fully paid` their loans or those who does not, so we will tansform de dependent variable into binary problem, to ones or ceros depending on that condition. 

In [3]:
    # We will only select those observations with "Fully Paid" & "Charged Off"
data_binary = data[(data['loan_status'] == "Fully Paid") | (data['loan_status'] == "Charged Off")]

    # Now, we will transform into 0 & 1
dummy_dict = {"Fully Paid":0, "Charged Off":1}

    # Finally, we use the dictiony in the dataset
data = data_binary.replace({"loan_status": dummy_dict})

### View data

Firstly, we are goint to take a look to our dataframe:

In [4]:
    # First 5 rows using html display in order to view all the columns
data = data
display(HTML(data.head().to_html()))

In [5]:
    # Last 5 rows using html display in order to view all the columns
data = data
display(HTML(data.tail().to_html()))

In [6]:
    # Data dimension
data.shape

(442567, 151)

### Data wrangling

After studying the whole dictionary, place in `references/dictionary.md`, we saw that there are some variables that we can not use them, because is information a posteriori, some examples of these columns are: `collection_recovery_fee`, `installment`, `last_credit_pull_d`, `last_pymnt_amnt`...

In [7]:
    # First, we make an objetc with the information of these columns
columns = ["collection_recovery_fee", "dti", "dti_joint", "fico_range_high", "fico_range_low", "grade", "initial_list_status", 
           "installment", "last_credit_pull_d", "last_fico_range_high", "last_fico_range_low", "last_pymnt_amnt", 
           "last_pymnt_d", "loan_amnt", "mths_since_last_delinq", "mths_since_last_major_derog", "mths_since_rcnt_il", 
           "next_pymnt_d", "mths_since_recent_bc", "mths_since_recent_bc_dlq", "mths_since_recent_inq", 
           "mths_since_recent_revol_delinq", "out_prncp", "out_prncp_inv", "pymnt_plan", "recoveries", "sub_grade", 
           "tot_coll_amt", "tot_cur_bal", "total_bal_ex_mort", "total_pymnt", "total_pymnt_inv", "total_rec_int", 
           "total_rec_late_fee", "total_rec_prncp", "total_rev_hi_lim", "verification_status", "verification_status_joint"]

    # Then we drop those columns
data = data.drop(columns = columns)

Once we have viewed our dataset, we continue exploring the column types and nulls...

In [8]:
data.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 442567 entries, 8 to 884875
Data columns (total 113 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   id                                          442567 non-null  int64  
 1   member_id                                   0 non-null       float64
 2   funded_amnt                                 442567 non-null  int64  
 3   funded_amnt_inv                             442567 non-null  float64
 4   term                                        442567 non-null  object 
 5   int_rate                                    442567 non-null  object 
 6   emp_title                                   417552 non-null  object 
 7   emp_length                                  418693 non-null  object 
 8   home_ownership                              442567 non-null  object 
 9   annual_inc                                  442567 non-null  float64


As we can see above we have plenty of columns with nulls, so we are going to drop those columns which more than a 75% of the information is null.

In [9]:
    # For this operation we will use drop.na(), but firstly we will create an objet with the scale
thresh = round(len(data) * 0.75) 

data = data.dropna(thresh = thresh, axis = 1)

    #Lets check the data shape
data.shape

(442567, 63)

We can see that now we have only 64 columns, instead of 151. This is a huge reduction of the variables, but we continue having too much, in order to reduce them, we will study which of them are important in order to predict `loan_status`.

Other method we can use in order to select those variables which are important is, we could use a correlation matrix and select those whith hight correlation values, but with 65 columns will be too difficult... so we will use a smarter way:

*Select those variables with more than 80% of correlation between them and drop them*

In [10]:
    # To start, we create our objet called "correlated_features" and our correlation parameter
correlated_features = set()
corr_param = 0.8

    # Them, we drop our predict variable
correlation_matrix = data.drop('loan_status', axis=1).corr()

    # Finally we use a loop
for i in range(len(correlation_matrix.columns)):
        for j in range(i):
            if abs(correlation_matrix.iloc[i, j]) > corr_param:
                colname = correlation_matrix.columns[i]
                correlated_features.add(colname)

correlated_features

{'funded_amnt_inv',
 'num_bc_sats',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_30dpd',
 'percent_bc_gt_75',
 'tot_hi_cred_lim',
 'total_bc_limit'}

In [11]:
data = data.drop(correlated_features, axis = 1)

Now we will drop those columns which are not important to predict `loan_status`, using our business knowlege, some examples of them are `url`, `policy_code`, `title`, `purpose`...

In [12]:
    # First, we make an objetc with the information of these columns
columns = ["emp_title", "id", "policy_code", "purpose", "title", "url", "zip_code"]

    # Then we drop those columns
data = data.drop(columns = columns)

    # Check the dataset
data.shape

(442567, 46)

***

***

**Vamos a ir viendo cuales son los valores únicos de cada variable e intentar pasarlas a factor, para poder practicar un RFE y un PCA, de manera que nos quedemos con las buenas...**

In [13]:
display(HTML(data.head().to_html()))

In [14]:
data['term'].unique()

array([' 36 months', ' 60 months'], dtype=object)

In [15]:
data['emp_length'].unique()

array(['6 years', '2 years', '7 years', '10+ years', '3 years',
       '< 1 year', '9 years', '4 years', '1 year', '5 years', nan,
       '8 years'], dtype=object)

In [16]:
data['home_ownership'].unique()

array(['OWN', 'RENT', 'MORTGAGE', 'ANY'], dtype=object)

In [17]:
data['addr_state'].unique()

array(['CA', 'WA', 'TX', 'MA', 'NJ', 'FL', 'CT', 'MT', 'MN', 'WI', 'VA',
       'GA', 'NY', 'DC', 'MD', 'PA', 'MO', 'SC', 'NV', 'MI', 'OH', 'CO',
       'LA', 'NM', 'TN', 'IL', 'KY', 'AZ', 'IN', 'OR', 'ME', 'NC', 'OK',
       'ND', 'RI', 'ID', 'AR', 'HI', 'AL', 'MS', 'UT', 'KS', 'VT', 'NE',
       'NH', 'SD', 'DE', 'AK', 'WY', 'WV', 'IA'], dtype=object)

In [18]:
data['application_type'].unique()

array(['Individual', 'Joint App'], dtype=object)

In [19]:
data['hardship_flag'].unique()

array(['N'], dtype=object)

In [20]:
data['disbursement_method'].unique()

array(['Cash', 'DirectPay'], dtype=object)

In [21]:
data['debt_settlement_flag'].unique()

array(['N', 'Y'], dtype=object)

**Primera conculsión, como solo tenemos un único tipo en `hardship_flag` la tenemos que eliminar, porque no aporta info**

**Vamos a revisar los tipos que tenemos de objetos que tenemos para ver qué cambios tenemos que hacer**

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 442567 entries, 8 to 884875
Data columns (total 46 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   funded_amnt                 442567 non-null  int64  
 1   term                        442567 non-null  object 
 2   int_rate                    442567 non-null  object 
 3   emp_length                  418693 non-null  object 
 4   home_ownership              442567 non-null  object 
 5   annual_inc                  442567 non-null  float64
 6   issue_d                     442567 non-null  object 
 7   loan_status                 442567 non-null  int64  
 8   addr_state                  442567 non-null  object 
 9   delinq_2yrs                 442567 non-null  int64  
 10  earliest_cr_line            442567 non-null  object 
 11  inq_last_6mths              442567 non-null  int64  
 12  open_acc                    442567 non-null  int64  
 13  pub_rec       

- `term`: Hay que quitarle "months" y pasarlo a factor, si es a 36 meses 0, si es a 60 1 **✓**

- `int_rate`: Eliminar el símbolo "%" y pasarlo a float

- `emp_length`: Factorizarlo

- `home_ownership`: Factorizarlo

- `issue_d`: Ver que hacemos con las fechas... OJO no sé si se refiere a la fecha en la que se solicitó o se concedió, si es la última hay que quitarla

- **`loan_status`**: Factorizarlo

- `addr_state`: Factorizarlo

- `earliest_cr_line`: Ver que hacemos con las fechas... Igual esta no interesa

- `revol_util`: Quitar el símbolo "%" y pasarlo a float

- `application_type`: Factorizarlo

- `hardship_flag`: Cargarsela porque solo tiene un tipo

- `disbursement_method`: Factorizarla

- `debt_settlement_flag`: Factorizarla

In [23]:
display(HTML(data.head().to_html()))

In [24]:
    # Pasamos a cateogry la variable
data["term"] = data["term"].astype('category')
    # Comprobamos
data.term.dtypes

CategoricalDtype(categories=[' 36 months', ' 60 months'], ordered=False)

In [25]:
    # Machacamos la columna con 1 y 0
data["term"] = data["term"].cat.codes

    # Comprobamos el resultado
display(HTML(data.head().to_html()))

**Una vez arreglemos todo lo de arriba podemos utilizar RFE para seleccionar las más importantes**